source/user/lmbench3/doc/bw_tcp.8
.\" $Id: bw_tcp.8,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $
.TH BW_TCP 1 "$Date: 2006-11-23 11:25:34 $" "(c)1994 Larry McVoy" "LMBENCH"
.SH NAME
bw_tcp \- time data movement through TCP/IP sockets
.SH SYNOPSIS
.B bw_tcp
[
.I "-m <message size>"
]
[
.I "-M <total bytes>"
]
[
.I "-P <parallelism>"
]
[
.I "-W <warmups>"
]
[
.I "-N <repetitions>"
]
.I "server"
.br or
.B bw_tcp
.I -s
.br or
.B bw_tcp
.I "-S <server>"
.SH DESCRIPTION
.B bw_tcp
is a client/server program that moves data over a TCP/IP socket.  Nothing is
done with the data on either side; 
.I "total bytes"
of data is moved in 
.I "message size"
chunks.
.LP
.B bw_tcp
has three forms of usage: as a server (-s), as a client (bw_tcp localhost), and
as a shutdown (bw_tcp -S localhost).
.LP
The default amount of data is 10MB.  The client form may specify a different
amount of data.  Specifications may end with ``k'' or ``m'' to mean
kilobytes (* 1024) or megabytes (* 1024 * 1024).
.SH OUTPUT
Output format is
.ft CB
Socket bandwidth using localhost: 2.32 MB/sec
.ft
.SH MEMORY UTILIZATION
This benchmark can move up to six times the requested memory per process
when run through the loopback device.
There are two processes, the sender and the receiver.
Most Unix
systems implement the read/write system calls as a bcopy from/to kernel space
to/from user space.  Bcopy will use 2-3 times as much memory bandwidth:
there is one read from the source and a write to the destionation.  The
write usually results in a cache line read and then a write back of
the cache line at some later point.  Memory utilization might be reduced
by 1/3 if the processor architecture implemented "load cache line"
and "store cache line" instructions (as well as getcachelinesize).
.SH ACKNOWLEDGEMENT
Funding for the development of
this tool was provided by Sun Microsystems Computer Corporation 
and Silicon Graphics, Inc.
.SH SEE ALSO
lmbench(8).
.SH "AUTHOR"
Carl Staelin and Larry McVoy
.PP
Comments, suggestions, and bug reports are always welcome.


source/History
Ralink AP SDK - History

Version 4.0.1.0
===============

- [IPv6] Avoid both eth2.2 and eth2 set same IPV6 address after receiving RA packet in gateway mode.
- [IPv6] Fix HNAT misbinding issue (RT6855/6)
- [IPv6] Fix ping fail(ping ::1) issue of loopback interface when we try to enable and disable disable_ipv6 parameter in /proc directory (RT6855/6)
- [Switch] Fix legacy switch(RT3052/RT3352/RT5350) settings vlan 12345 error
- [PPPoE] Kernel panic after PAP fail (RT6855/6)
- [PPTP] Kernel panic after PPTP peer terminated the link (RT6855/6)
- [SPI Flash] There are some noise if we try to program flash while I2S is playing music (RT3052/RT3883/RT3352/RT5350)
- [GPIO] Fix GPIO interrupt handler cannot be trigged correctly (RT6855/6)
- [GUI] Fix WPS bug for test plan
- [NET] Recover irqs_disabled() check under SKB Recycle to fix warning messages, but it decreases RT6855/6's maximum peek throughput
- [Uboot] RT6855/6 does watchdog reset at cold start to fix timer interrupt issue

Version 4.0.0.0
===============

PLATFORM:

- support RT3052/RT3883/RT3352/RT5350 (linux-2.6.21.x)
- support RT6855/RT6856 (linux-2.6.36MT.x)

NEW FUNCTIONS:

- [IPv6 ready logo Phase II] RT6855/RT6856
- [IPv6] support IPv6 Rapid Deployment
- [IPv6] support IPv6 DS-Lite
- [GUI] support two giga phy port
- [GUI] display AP Client site_survey

IMPROVEMENT:

- [IPv6] support ipv6 tftp server
- [PPTP/L2TP] performace tunning at Linux 2.6.21 (PPTP improve 25%, L2TP improve 10% compare to SDK 3.6)
- [SAMBA] significant performance fine tune on full loading scenario (dual-band concurrent AP with USB access)
- [NAND Flash] config page size and address cycle based on chip mode
- [I2S] Add MMAP for user/kernel
- [GUI] support WPS 2.0 

Bug fixed:

- [Kernel NVRAM] memory leakage / initiate failed / free cache / wirte data exceed 1024
- [RAETH] fix STP drop issue on RT3052 chipst ONLY
- [RAETH] fix switch abnormal after ifconfig down&up
- [RAETH] fix PDMA hang issue under heavy load test
- [RAETH] fix TSO bug if USB storage is enabled
- [USB device] system will hang whiling loading gadget driver. Turning off power-saving on both port0 and port1
- [USB device] force host to do re-enumeration to solve "unrecognize device" when re-plug-in in windows platform.
- [IPv6] kernel panic bug if enable ipv6 privacy extension or router prereference
- [IPv6] fix NS packet loss for ipv6 DAD
- [GUI] flush conntrack table while restart/deleting rules
- [NAND Flash] fix nand flash not work for MTD_ANY mode
- [IGMPPROXY] fixed the overflow FATAL bug while handling bit47=1 mac address
- [GPIO] fixed RT3352/RT5350 extended GPIO pins mapping



source/user/lmbench3/src/bw_tcp.c
/*
 * bw_tcp.c - simple TCP bandwidth test
 *
 * Three programs in one -
 *	server usage:	bw_tcp -s
 *	client usage:	bw_tcp [-m <message size>] [-M <total bytes>] [-P <parallelism>] [-W <warmup>] [-N <repetitions>] hostname 
 *	shutdown:	bw_tcp -hostname
 *
 * Copyright (c) 2000 Carl Staelin.
 * Copyright (c) 1994 Larry McVoy.  Distributed under the FSF GPL with
 * additional restriction that results may published only if
 * (1) the benchmark is unmodified, and
 * (2) the version in the sccsid below is included in the report.
 * Support for this development by Sun Microsystems is gratefully acknowledged.
 */
char	*id = "$Id: bw_tcp.c,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $\n";
#include "bench.h"

typedef struct _state {
	int	sock;
	uint64	move;
	int	msize;
	char	*server;
	int	fd;
	char	*buf;
} state_t;

void	server_main();
void	client_main(int parallel, state_t *state);
void	source(int data);

void	initialize(iter_t iterations, void* cookie);
void	loop_transfer(iter_t iterations, void *cookie);
void	cleanup(iter_t iterations, void* cookie);

int
main(int ac, char **av)
{
	int	parallel = 1;
	int	warmup = LONGER;
	int	repetitions = TRIES;
	int	shutdown = 0;
	state_t state;
	char	*usage = "-s\n OR [-m <message size>] [-M <bytes to move>] [-P <parallelism>] [-W <warmup>] [-N <repetitions>] server\n OR -S serverhost\n";
	int	c;
	
	state.msize = 0;
	state.move = 0;

	/* Rest is client argument processing */
	while (( c = getopt(ac, av, "sS:m:M:P:W:N:")) != EOF) {
		switch(c) {
		case 's': /* Server */
			if (fork() == 0) {
				server_main();
			}
			exit(0);
			break;
		case 'S': /* shutdown serverhost */
		{
			int	conn;
			conn = tcp_connect(optarg, TCP_DATA, SOCKOPT_NONE);
			write(conn, "0", 1);
			exit(0);
		}
		case 'm':
			state.msize = bytes(optarg);
			break;
		case 'M':
			state.move = bytes(optarg);
			break;
		case 'P':
			parallel = atoi(optarg);
			if (parallel <= 0) lmbench_usage(ac, av, usage);
			break;
		case 'W':
			warmup = atoi(optarg);
			break;
		case 'N':
			repetitions = atoi(optarg);
			break;
		default:
			lmbench_usage(ac, av, usage);
			break;
		}
	}

	if (optind < ac - 2 || optind >= ac) {
		lmbench_usage(ac, av, usage);
	}

	state.server = av[optind++];

	if (state.msize == 0 && state.move == 0) {
		state.msize = state.move = XFERSIZE;
	} else if (state.msize == 0) {
		state.msize = state.move;
	} else if (state.move == 0) {
		state.move = state.msize;
	}

	/* make the number of bytes to move a multiple of the message size */
	if (state.move % state.msize) {
		state.move += state.msize - state.move % state.msize;
	}

	/*
	 * Default is to warmup the connection for seven seconds, 
	 * then measure performance over each timing interval.
	 * This minimizes the effect of opening and initializing TCP 
	 * connections.
	 */
	benchmp(initialize, loop_transfer, cleanup, 
		0, parallel, warmup, repetitions, &state);
	if (gettime() > 0) {
		fprintf(stderr, "%.6f ", state.msize / (1000. * 1000.));
		mb(state.move * get_n() * parallel);
	}
}

void
initialize(iter_t iterations, void *cookie)
{
	int	c;
	char	buf[100];
	state_t *state = (state_t *) cookie;

	if (iterations) return;

	state->buf = valloc(state->msize);
	if (!state->buf) {
		perror("valloc");
		exit(1);
	}
	touch(state->buf, state->msize);

	state->sock = tcp_connect(state->server, TCP_DATA, SOCKOPT_READ|SOCKOPT_WRITE|SOCKOPT_REUSE);
	if (state->sock < 0) {
		perror("socket connection");
		exit(1);
	}
	sprintf(buf, "%lu", state->msize);
	if (write(state->sock, buf, strlen(buf) + 1) != strlen(buf) + 1) {
		perror("control write");
		exit(1);
	}
}

void 
loop_transfer(iter_t iterations, void *cookie)
{
	int	c;
	uint64	todo;
	state_t *state = (state_t *) cookie;

	while (iterations-- > 0) {
		for (todo = state->move; todo > 0; todo -= c) {
			if ((c = read(state->sock, state->buf, state->msize)) <= 0) {
				exit(1);
			}
			if (c > todo) c = todo;
		}
	}
}

void
cleanup(iter_t iterations, void* cookie)
{
	state_t *state = (state_t *) cookie;

	if (iterations) return;

	/* close connection */
	(void)close(state->sock);
}

void
server_main()
{
	int	data, newdata;

	GO_AWAY;

	data = tcp_server(TCP_DATA, SOCKOPT_WRITE|SOCKOPT_REUSE);
	if (data < 0) {
		perror("server socket creation");
		exit(1);
	}

	signal(SIGCHLD, sigchld_wait_handler);
	for ( ;; ) {
		newdata = tcp_accept(data, SOCKOPT_WRITE);
		switch (fork()) {
		    case -1:
			perror("fork");
			break;
		    case 0:
			source(newdata);
			exit(0);
		    default:
			close(newdata);
			break;
		}
	}
}

/*
 * Read the message size.  Keep transferring 
 * data in message-size sized packets until
 * the socket goes away.
 */
void
source(int data)
{
	size_t	count, m;
	unsigned long	nbytes;
	char	*buf, scratch[100];

	/*
	 * read the message size
	 */
	bzero(scratch, 100);
	if (read(data, scratch, 100) <= 0) {
		perror("control nbytes");
		exit(7);
	}
	sscanf(scratch, "%lu", &nbytes);
	m = nbytes;

	/*
	 * A hack to allow turning off the absorb daemon.
	 */
     	if (m == 0) {
		tcp_done(TCP_DATA);
		kill(getppid(), SIGTERM);
		exit(0);
	}

	buf = valloc(m);
	bzero(buf, m);

	/*
	 * Keep sending messages until the connection is closed
	 */
	while (write(data, buf, m) == m) {
#ifdef	TOUCH
		touch(buf, m);
#endif
	}
	free(buf);
}
